File: java/src/test/java/md/remit/A2ATest.java
package md.remit;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import java.util.List;
import java.util.Map;

import static org.assertj.core.api.Assertions.*;

@DisplayName("A2A types and helper tests")
class A2ATest {

    // ─── Record construction ──────────────────────────────────────────────────

    @Test
    @DisplayName("Extension record holds fields")
    void testExtension() {
        A2A.Extension ext = new A2A.Extension("https://remit.md/x402", "x402 payments", true);
        assertThat(ext.uri()).isEqualTo("https://remit.md/x402");
        assertThat(ext.description()).isEqualTo("x402 payments");
        assertThat(ext.required()).isTrue();
    }

    @Test
    @DisplayName("Capabilities record holds fields")
    void testCapabilities() {
        A2A.Capabilities cap = new A2A.Capabilities(true, false, true, List.of());
        assertThat(cap.streaming()).isTrue();
        assertThat(cap.pushNotifications()).isFalse();
        assertThat(cap.stateTransitionHistory()).isTrue();
    }

    @Test
    @DisplayName("Skill record holds fields")
    void testSkill() {
        A2A.Skill skill = new A2A.Skill("pay", "Pay", "Send USDC", List.of("payment"));
        assertThat(skill.id()).isEqualTo("pay");
        assertThat(skill.name()).isEqualTo("Pay");
        assertThat(skill.tags()).containsExactly("payment");
    }

    @Test
    @DisplayName("Fees record holds fields")
    void testFees() {
        A2A.Fees fees = new A2A.Fees(100, 50, 10000);
        assertThat(fees.standardBps()).isEqualTo(100);
        assertThat(fees.preferredBps()).isEqualTo(50);
        assertThat(fees.cliffUsd()).isEqualTo(10000);
    }

    @Test
    @DisplayName("TaskStatus record holds fields")
    void testTaskStatus() {
        A2A.TaskStatus status = new A2A.TaskStatus("completed", null);
        assertThat(status.state()).isEqualTo("completed");
        assertThat(status.message()).isNull();
    }

    @Test
    @DisplayName("ArtifactPart record holds data")
    void testArtifactPart() {
        A2A.ArtifactPart part = new A2A.ArtifactPart("data", Map.of("txHash", "0xabc"));
        assertThat(part.kind()).isEqualTo("data");
        assertThat(part.data()).containsKey("txHash");
    }

    @Test
    @DisplayName("Task record holds fields")
    void testTask() {
        A2A.Task task = new A2A.Task(
            "task-1",
            new A2A.TaskStatus("completed", null),
            List.of()
        );
        assertThat(task.id()).isEqualTo("task-1");
        assertThat(task.status().state()).isEqualTo("completed");
    }

    // ─── getTaskTxHash ────────────────────────────────────────────────────────

    @Test
    @DisplayName("getTaskTxHash extracts txHash from artifacts")
    void testGetTaskTxHashFound() {
        A2A.Task task = new A2A.Task(
            "task-1",
            new A2A.TaskStatus("completed", null),
            List.of(new A2A.Artifact("result", List.of(
                new A2A.ArtifactPart("data", Map.of("txHash", "0xabc123"))
            )))
        );
        assertThat(A2A.getTaskTxHash(task)).isEqualTo("0xabc123");
    }

    @Test
    @DisplayName("getTaskTxHash returns null when no txHash")
    void testGetTaskTxHashNotFound() {
        A2A.Task task = new A2A.Task(
            "task-2",
            new A2A.TaskStatus("completed", null),
            List.of(new A2A.Artifact("result", List.of(
                new A2A.ArtifactPart("data", Map.of("other", "value"))
            )))
        );
        assertThat(A2A.getTaskTxHash(task)).isNull();
    }

    @Test
    @DisplayName("getTaskTxHash returns null for empty artifacts")
    void testGetTaskTxHashEmpty() {
        A2A.Task task = new A2A.Task(
            "task-3",
            new A2A.TaskStatus("completed", null),
            List.of()
        );
        assertThat(A2A.getTaskTxHash(task)).isNull();
    }

    @Test
    @DisplayName("getTaskTxHash returns null for null artifacts")
    void testGetTaskTxHashNull() {
        A2A.Task task = new A2A.Task(
            "task-4",
            new A2A.TaskStatus("completed", null),
            null
        );
        assertThat(A2A.getTaskTxHash(task)).isNull();
    }

    // ─── SendOptions ──────────────────────────────────────────────────────────

    @Test
    @DisplayName("SendOptions two-arg constructor")
    void testSendOptionsTwoArg() {
        A2A.SendOptions opts = new A2A.SendOptions("0xRecipient", 1.50);
        assertThat(opts.to()).isEqualTo("0xRecipient");
        assertThat(opts.amount()).isEqualTo(1.50);
        assertThat(opts.memo()).isEmpty();
        assertThat(opts.mandate()).isNull();
    }

    @Test
    @DisplayName("SendOptions three-arg constructor")
    void testSendOptionsThreeArg() {
        A2A.SendOptions opts = new A2A.SendOptions("0xRecipient", 2.00, "test memo");
        assertThat(opts.to()).isEqualTo("0xRecipient");
        assertThat(opts.amount()).isEqualTo(2.00);
        assertThat(opts.memo()).isEqualTo("test memo");
    }

    @Test
    @DisplayName("SendOptions full constructor with mandate")
    void testSendOptionsWithMandate() {
        A2A.IntentMandate mandate = new A2A.IntentMandate(
            "m-1", "2026-12-31", "0xissuer", Map.of("maxAmount", "100", "currency", "USDC")
        );
        A2A.SendOptions opts = new A2A.SendOptions("0xRecipient", 5.00, "with mandate", mandate);
        assertThat(opts.mandate()).isNotNull();
        assertThat(opts.mandate().mandateId()).isEqualTo("m-1");
    }

    // ─── IntentMandate ────────────────────────────────────────────────────────

    @Test
    @DisplayName("IntentMandate holds fields")
    void testIntentMandate() {
        A2A.IntentMandate m = new A2A.IntentMandate(
            "m-123", "2026-12-31T23:59:59Z", "0xIssuer",
            Map.of("maxAmount", "1000", "currency", "USDC")
        );
        assertThat(m.mandateId()).isEqualTo("m-123");
        assertThat(m.expiresAt()).isEqualTo("2026-12-31T23:59:59Z");
        assertThat(m.issuer()).isEqualTo("0xIssuer");
        assertThat(m.allowance()).containsKeys("maxAmount", "currency");
    }

    // ─── Client construction ──────────────────────────────────────────────────

    @Test
    @DisplayName("Client constructor with valid endpoint")
    void testClientConstruction() {
        // We cannot test real RPC calls, but we can verify construction succeeds.
        // Use a mock signer.
        md.remit.signer.Signer signer = new md.remit.signer.Signer() {
            @Override
            public byte[] sign(byte[] digest) { return new byte[65]; }
            @Override
            public String address() { return "0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266"; }
        };
        A2A.Client client = new A2A.Client("https://remit.md/a2a", signer, 8453, null);
        assertThat(client).isNotNull();
    }

    @Test
    @DisplayName("Client.fromCard with AgentCard")
    void testClientFromCard() {
        md.remit.signer.Signer signer = new md.remit.signer.Signer() {
            @Override
            public byte[] sign(byte[] digest) { return new byte[65]; }
            @Override
            public String address() { return "0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266"; }
        };
        A2A.AgentCard card = new A2A.AgentCard(
            "0.1", "Remit", "Payment agent", "https://remit.md/a2a", "1.0",
            "https://remit.md/docs", null, null, null
        );
        A2A.Client client = A2A.Client.fromCard(card, signer);
        assertThat(client).isNotNull();
    }
}

File: go/a2a_test.go
package remitmd_test

import (
	"testing"

	remitmd "github.com/remit-md/sdk/go"
)

// Test key - well-known Hardhat/Foundry test key #0 (not a real wallet).
const testKey = "ac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80"

func TestNewA2AClient_ValidEndpoint(t *testing.T) {
	signer, err := remitmd.NewPrivateKeySigner(testKey)
	if err != nil {
		t.Fatalf("signer: %v", err)
	}
	client, err := remitmd.NewA2AClient(remitmd.A2AClientOptions{
		Endpoint: "https://remit.md/a2a",
		Signer:   signer,
		ChainID:  8453,
	})
	if err != nil {
		t.Fatalf("NewA2AClient failed: %v", err)
	}
	if client == nil {
		t.Fatal("expected non-nil client")
	}
}

func TestNewA2AClient_InvalidEndpoint(t *testing.T) {
	signer, _ := remitmd.NewPrivateKeySigner(testKey)
	_, err := remitmd.NewA2AClient(remitmd.A2AClientOptions{
		Endpoint: "://invalid",
		Signer:   signer,
		ChainID:  8453,
	})
	if err == nil {
		t.Fatal("expected error for invalid endpoint")
	}
}

func TestA2AClientFromCard_DefaultChain(t *testing.T) {
	signer, _ := remitmd.NewPrivateKeySigner(testKey)
	card := &remitmd.AgentCard{
		URL: "https://remit.md/a2a",
	}
	client, err := remitmd.A2AClientFromCard(card, signer)
	if err != nil {
		t.Fatalf("A2AClientFromCard failed: %v", err)
	}
	if client == nil {
		t.Fatal("expected non-nil client")
	}
}

func TestA2AClientFromCard_BaseSepolia(t *testing.T) {
	signer, _ := remitmd.NewPrivateKeySigner(testKey)
	card := &remitmd.AgentCard{
		URL: "https://testnet.remit.md/a2a",
	}
	client, err := remitmd.A2AClientFromCard(card, signer, "base-sepolia")
	if err != nil {
		t.Fatalf("A2AClientFromCard failed: %v", err)
	}
	if client == nil {
		t.Fatal("expected non-nil client")
	}
}

func TestA2AClientFromCard_Localhost(t *testing.T) {
	signer, _ := remitmd.NewPrivateKeySigner(testKey)
	card := &remitmd.AgentCard{URL: "http://localhost:3000/a2a"}
	client, err := remitmd.A2AClientFromCard(card, signer, "localhost")
	if err != nil {
		t.Fatalf("A2AClientFromCard failed: %v", err)
	}
	if client == nil {
		t.Fatal("expected non-nil client")
	}
}

func TestGetTaskTxHash_Found(t *testing.T) {
	task := &remitmd.A2ATask{
		ID: "test-task-1",
		Status: remitmd.A2ATaskStatus{
			State: "completed",
		},
		Artifacts: []remitmd.A2AArtifact{
			{
				Name: "result",
				Parts: []remitmd.A2AArtifactPart{
					{
						Kind: "data",
						Data: map[string]any{
							"txHash": "0xabc123",
						},
					},
				},
			},
		},
	}
	hash := remitmd.GetTaskTxHash(task)
	if hash != "0xabc123" {
		t.Errorf("expected 0xabc123, got %s", hash)
	}
}

func TestGetTaskTxHash_NotFound(t *testing.T) {
	task := &remitmd.A2ATask{
		ID: "test-task-2",
		Status: remitmd.A2ATaskStatus{
			State: "completed",
		},
		Artifacts: []remitmd.A2AArtifact{
			{
				Name: "result",
				Parts: []remitmd.A2AArtifactPart{
					{Kind: "data", Data: map[string]any{"other": "value"}},
				},
			},
		},
	}
	hash := remitmd.GetTaskTxHash(task)
	if hash != "" {
		t.Errorf("expected empty string, got %s", hash)
	}
}

func TestGetTaskTxHash_EmptyArtifacts(t *testing.T) {
	task := &remitmd.A2ATask{
		ID:        "test-task-3",
		Artifacts: nil,
	}
	hash := remitmd.GetTaskTxHash(task)
	if hash != "" {
		t.Errorf("expected empty string for nil artifacts, got %s", hash)
	}
}

func TestA2ATypes_Construction(t *testing.T) {
	ext := remitmd.A2AExtension{
		URI:         "https://remit.md/x402",
		Description: "x402 payments",
		Required:    true,
	}
	if ext.URI != "https://remit.md/x402" {
		t.Errorf("expected URI, got %s", ext.URI)
	}

	cap := remitmd.A2ACapabilities{Streaming: true}
	if !cap.Streaming {
		t.Error("expected streaming true")
	}

	skill := remitmd.A2ASkill{ID: "pay", Name: "Pay", Tags: []string{"payment"}}
	if skill.ID != "pay" {
		t.Errorf("expected ID pay, got %s", skill.ID)
	}
}

func TestIntentMandate_Fields(t *testing.T) {
	mandate := remitmd.IntentMandate{
		MandateID: "m-123",
		ExpiresAt: "2026-12-31T23:59:59Z",
		Issuer:    "0xabc",
	}
	if mandate.MandateID != "m-123" {
		t.Errorf("expected m-123, got %s", mandate.MandateID)
	}
}

func TestSendOptions_Fields(t *testing.T) {
	opts := remitmd.SendOptions{
		To:     "0xrecipient",
		Amount: 1.50,
		Memo:   "test",
	}
	if opts.To != "0xrecipient" {
		t.Errorf("expected 0xrecipient, got %s", opts.To)
	}
	if opts.Amount != 1.50 {
		t.Errorf("expected 1.50, got %f", opts.Amount)
	}
}
